Playtime Tracking

Kyep opened the pull request:
Our current jobs unlock system isn't great, for several reasons:
- No matter how active you are, you can't unlock jobs any faster. 
- Conversely, no matter how inactive you are, come back in a month and all jobs will be unlocked automatically.
- This results in players who have older accounts, but are completely new to Paradise, taking jobs like HoS and then trying to act how they do on their home severs - which does not work well here.
- It also results in new players realizing there is nothing they can do to unlock the jobs they really want any faster, so they tend to "go play another game for awhile" - and often don't come back. 

This PR:
- Enables the tracking of playtime per player. Disabled by default.
- Adds a new playtime-based job unlock system. Disabled by default. When enabled, heads-of-staff jobs require a few hours of playtime on Paradise to unlock. This prevents people from other servers jumping straight into head jobs on Paradise with no idea how they work on our server.

Please note:
- This system relies upon the database. It won't work with filesystem-based char profiles.
- The system can run in tracking-only mode. We can run it this way for a month, before enabling its job unlock system, so that _regular players will already have suitable jobs unlocked on day one_. EDIT: there is also a SQL query that can be run after patching, to grandfather all existing, active players, and make sure they're counted as having all jobs unlocked.
- This is not necessarily intended to fully replace the system of unlocking jobs based on your account being X days old. I'd would be _very happy_ if it did, but my more modest goal is to apply it to at least heads-of-staff level jobs. The idea being that people who've never played here before shouldn't be able to hop right into jobs like that, regardless of their account age. They should have to play a little on Paradise before taking jobs with such power/authority/responsibility.

The stats tracking added in this PR could be useful for things other than gating job access. They could also help detect people who cryo every round they don't get antag, for example. It could also be used to identify new players with old accounts.

**The main benefit, though, is the new gating system - basing access to jobs on experience, rather than simply account age.** This measures playtime far more accurately, and should dramatically reduce the number of cases of people playing Head roles without understanding the department they've signed up to manage.

**POST-PATCH INSTALL INSTRUCTIONS**

1) Add the col to the player table:
ALTER TABLE `player` ADD `exp` mediumtext NOT NULL DEFAULT ''

2) (Optional) Grandfather all active players, so they won't lose any jobs if the unlock system is enabled:
UPDATE player SET exp='Exempt=1' WHERE lastseen >= '2016-04-01'

3) Copy the USE_EXP_TRACKING section from config/example/config.txt to config/config.txt
Be sure to enable the USE_EXP_TRACKING directive.

:cl: Kyep
add: Added tracking for how much playtime each player has.
add: Added the ability to lock jobs based on playtime. E.g: must play for 10 hours to unlock HoS.
/:cl:

Twinmold93 commented:
A system that is based on the time people having playing the game, rather than just logging in and then not playing just to get full access is a smart move, especially in pushing towards a more reliable set of players for positions whom are expected to know the role, actually giving them time to learn how things on Paradise are done (like SOP, Space Law, and general demeanor) vs pulling their behavior from other servers and expecting nothing to come of it, for people to be the same way.

We have a level of roleplay and competence, and knowing a role well, through playing the lesser roles first, definitely helps us make sure they have time to learn the role first through observation, rather than just doing.

With a system like this, positions can also open up to players far sooner than before, if they show the aptitude to play the game, and the drive to do so, instead of waiting for the day counter to be up. This would actually prove beneficial to players of other servers who know how to do a role, and focus on a role to unlock it in a timely fashion, but theoretically initially focus in a department to do so, which can likely lead to players being better at something singular than just moderate at different roles.

This would also, very likely, make it so someone who focuses RD or CMO doesn't suddenly get HoP or HoS, but has no actually knowledge of how to do the position, and if they only intend to focus one or two departments, don't get thrown into departments they've not yet played as an expected head role from the positions they normally get getting taken by due to RNG. And if they are in a department they've never played, it'd be in a lesser role than the big one, expected to know how to do everything when they know nothing. This means our players in general are better protected, and people looking to ask questions of their boss on how to do things have a much better chance of getting the pointers they actually need than a 'I don't know' response.

FreeStylaLT commented:
Feels like more of an administration thing here I feel like

Fox-McCloud commented:
>Feels more of an administration thing here I feel like

TheBeoni commented:
I do not play too often which would lead to me not having unlocked security and forcing me to play something i do not want to so i can get something i want. Just for this aspect i am saying i do not like it.
On the other hand having HoS that is not completely shit would be awesome. And sec officers that actualy know space law.

tigercat2000 commented:
I really think this is completely unnecessary, and can be infinitely better enforced by administrators doing their job and monitoring for shitty heads/security.

theColdflame commented:
:+1: 

This is a needed change- time since first join is a bad metric and there's been a general increasein bad command players recently.

Necaladun commented:
I'm not 100% sure how we'd implement the locking, or if we'd turn it on at all.

The logging though, please yes so much.

TullyBurnalot commented:
@Necaladun @TheBeoni I may be wrong, but I believe that Kyets plan is, if this gets merged, to just let it run in the background for about a month or so without locking anything, to give regulars/semi-regulars plenty of time to get enough time logged on to unlock stuff.

TullyBurnalot commented:
@tigercat2000 No, not really. Our current job unlock system is ridiculously faulty. This at least offers a viable solution.

Fox-McCloud commented:
Quite frankly, we don't need an additional system to track which jobs are the most popular; we already have a perfectly viable system that does that, as is..and we don't use it.

Stats tracking is good, but there's no point in cluttering up the code with additional systems that we will likely never use (especially when an existing system is already in place for logging). This has come up before with PDAs and several other things. Invariably the data just sat there and was completely unused, whereupon we yanked said logging out at a later point.

We already don't use the data available to us, this won't be any different.

tigercat2000 commented:
@TullyBurnalot The concept isn't about player experience. It's about preventing low-effort griefing. 
A better system that isn't also ridiculously overcomplicated and tries to instill a forced "experience with role" system would, perhaps, be tracking how many times a player logs in (one per restart)

Spacemanspark commented:
:+1: I've been requesting this for a long, long time.

monster860 commented on code/modules/mob/new_player/new_player.dm:
```diff
@@ -246,6 +246,7 @@
 	if(!is_job_whitelisted(src, rank))	 return 0
 	if(!job.player_old_enough(src.client))	return 0
 	if(job.admin_only && !(check_rights(R_EVENT, 0))) return 0
+	if(!has_exp_for_job(src, rank)) return 0
```
Do not put the statement on the same lien as the if statement.

monster860 commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,154 @@
+		return -1
+
+	for(var/client/C in clients)
+		if(C.inactivity < (10 * 10 * 60))
```
Use the `MINUTES` macro instead, like this: `10 MINUTES`

TullyBurnalot commented:
@tigercat2000 So I can log in. Then either sit there doing nothing. And that would count.

monster860 commented:
@TullyBurnalot Actually, this pr doesn't count time when you are inactive.

tigercat2000 commented:
@TullyBurnalot that **still** fucking prevents low-effort griefing
you have to dedicate resources to idle constantly for god knows how long

Davdichan commented:
In theory, I can understand the Captain requirement needing every Command job unlocked first.

In practice, it's dumb too restrictive and sends the wrong the message to Captains on how they should manage the crisis.

Captain job should be unlocked by x hours of play across command in total, a sum that can be reached by playing one job or all 5 (8 if you count karma jobs)

Captain should never feel like the most skilled person on the station, but requiring Captain to have experience in all departments is going to reinforce that feeling. The Captain's job is to command. If there is a problem, direct the departments to fix it, and lacking a Head of Department, find someone capable to do the job. Engine not set up? Ride the CE's ass or have the HoP hire new engineers from other personnel, or maybe ask the RD if any Engi Cyborgs have been made that can wire the solars.

The Captain should never feel "well I can do it myself faster than I can ask someone else to do it." It's bad play, power gaming, and leads to the Captain often nicking gear from various departments to make themselves Super Captains without any real cause or justification for any of it.

tl;dr a good Captain doesn't need to know how to RnD, set up an engine, run cargo or how to do surgery or the various niche parts of Space Law and SoP for every department. A good Captain just needs to know these things exist and where to find people to do these jobs. I.E. COMMAND.

Kyep commented:
@Davdichan 
Right now, the only requirement in this system to unlock Captain is 240 minutes total playtime in any combination of Command jobs. That can be one two-hour shift as CE. One two-hour shift as RD. An hour as CMO, an hour as HoP. Any of these would qualify. To be clear: *there is no requirement to unlock every other head job before you unlock Captain*. That is not in this PR, and never has been.

Vivalas commented:
2h != 240 mins by the way.

I like this, especially since it puts emphasis on actually, you know, playing the game. So if you want to actually be active and try to learn the game you can, and get to have a job with responsibility.

Necaladun commented:
If this is linked to an adminverb I see it as a very useful tool for just getting info on a player quickly.

Whether to lock jobs behind it or not is an administrative thing to be discussed elsewhere, the code itself here seems good and useful.

Earthdivine commented:
I'd much rather not have Job tracking, unless it's not horribly restrictive in time compared to some of our CURRENT ones (30 days for Antag, seriously?).

But time based tracking entirely something we need. That will keep people from making multiple accounts and having a plethora of accounts open to them to grief with. My only problem are people who are coming back from breaks in the server and suddenly having jobs locked out.

IcyV commented:
I am absolutely in love with the idea of all this being logged. Locking a lot of the jobs behind random timers is...questionable at best to me. Botany has arguably one of the highest potentials for mass-grief, for example. An officer might annoy a few people if they're especially terrible, botany could end a round. 

I like the concept of something like the Captain or specific Head roles needing a barrier of entry given they can massively influence a round by doing absolutely nothing. But randomly locking other jobs out just feels like it'll kill off potential interest in the server. Why would I want to play somewhere that has everything I consider interesting locked out to me for a long time? It keeps closing in the community to smaller and smaller sets of people. Not that I'm saying the current system is any better.

>but my more modest goal is to apply it to at least heads-of-staff level jobs

I think that's honestly the best use of it, at least at first. Any random schmuck can get a quick transfer or get the materials to grief or be terrible easily enough from pretty much any job but head roles have an extra layer of responsibility and requirement to them. A scientist going and shooting people up with syringes and throwing grenades is a nuisance in the round. Losing the RD is potentially crippling.

Kyep commented:
@Earthdivine @IcyV @Necaladun @TullyBurnalot @Vivalas @Spacemanspark @Fox-McCloud @theColdflame @Twinmold93 @FreeStylaLT @monster860 @Davdichan

Changes made today in Update 2:
- Player Panel now shows general EXP for all players under their CKEY. Clicking this will show a very detailed EXP report for that player. This report includes all their XP totals in all categories, their immunity/grandfather status (see later), a list of every restricted job they've unlocked, and a list of every restricted job they have NOT unlocked (with their progression towards unlocking it). Players also get a verb that shows them the same info for their own account, under Special Verbs->EXP.
- EXP is now granted automatically every minute, rather than at the end of rounds. This means if you leave part-way through a round, you still get credit for the time you were in the round.
- Mentors now have an "Check Player XP" verb (under Admin tab) that shows general EXP (ie: relative newness to Paradise) for all connected players.
- Immunity is now supported. Simply run the SQL "UPDATE player SET exp='imm=1' WHERE lastseen >= '2016-04-01'" after merging, and *everyone who'se played on Paradise recently will have all jobs unlocked automatically.* The unlocking system will apply only to new players. 
- Split the config flags into use_exp_restrictions_heads and use_exp_restrictions_other. The former applies only to heads/AI, the latter applies to all jobs. This system can be run in tracking-only mode, in restricts-heads-only mode, or in applies-to-most-jobs mode. It is up to you. The suggested configuration is tracking on, heads on, other jobs off, with the SQL applied so that anyone who has played here recently bypasses the time requirement.
- Changed the timers on jobs to be something realistic, not based on test data. Broadly speaking: Captain requires 10h of command job XP, other heads require 10h of XP in their department, AI requires 5h of silicon XP, warden/IAA require 5 hours of security EXP, and MD/Chemist/Genetecist/Viro/Scientist/Robotocist/Detective/SecOfficer/Borg require 1-5 hours of general EXP (the kind you earn by playing anything at all except ghost). Even with non-head job progression enabled, the great majority of jobs unlock after only a few hours of play. The only jobs with remotely demanding unlock requirements are the head jobs.

Overall:
- Existing players won't be locked out of any jobs.
- New players, however, will have to spend some time in a department before playing as the head of that department.
- The system is highly configurable. You can run it however you like.

Necaladun commented:
Don't worry about any of the specific timers for now, we'll want to tweak that a LOT. Get the tracking any viewing via adminverbs in first, then we can look to see how to lock things.

Fox-McCloud commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,237 @@
+/client/proc/cmd_mentor_check_player_exp()	//Allows mentors / admins to determine who the newer players are.
+	set category = "Admin"
+	set name = "Check Player XP"
+	if(!check_rights(R_MENTOR|R_MOD|R_ADMIN))
```
Mentors should absolutely not have this, at all. They're not meant to have any tools other than their ability to answer mentorhelps. They're  a mentor, not administration.

Kyep replied:
cmd_mentor_check_new_players is an existing, mentor-accessible command which shows mentors who the new players are on the server, based on # of days since account creation.

cmd_mentor_check_player_exp is the exact same thing, just based on XP/playtime instead of days since registration. I can make it admin-only if that's what people want, but personally, a "show me people new to the server" command seems like exactly the sort of thing that would help Mentors identify and help new players.

Kyep commented:
Changes in Update 3:
- Hid XP info from mentors (Fox request)
- Commented the verb that lets normal players see their XP (Neca request)
- Added logging for usage of admin XP verbs
- Deleted old commented code
- Formatted XP numbers in reports as "1h", "3h", "27h" etc rather than a # of points/minutes.

Kyep commented:
@Fox-McCloud @TheDZD @Necaladun 
- Requesting "Merge Conflict" & "Work In Progress" labels removed, "Config Change", "SQL Change", and "Feature" labels added. 
- Requesting review. I'm not sure what else I'd add to this PR to make it better. Still up for making fixes if issues are found, but I don't have any further features in mind for this.

Ty-Omaha commented:
After having countless CMOs have no idea what they're doing this NEEDS to be in pls

/endsalt

Iamgoofball commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,234 @@
+				else
+					play_records[rtype] = text2num(play_records[rtype])
+
+			if(C.mob.stat == CONSCIOUS && C.mob.mind.assigned_role)
```
:-1:

if you die you shouldn't lose the experience, fuck off with this shit its about learnning to play not whether you lost those 2 hours of experience in a freak accident

Kyep replied:
Dying does not remove previously accumulated experience.

It simply means that, while dead, your time counts as ghost time, not normal playing time. 

And that is assuming you're not defibbed/cloned/etc.

Iamgoofball replied:
Fuck me I misread this shit

Fox-McCloud commented:
Was this tested? DId you set up a SQL? Did you test it out after SQL was set up and how the various interactions with it work?

Kyep commented:
Yes, I tested it. Yes, I tested it with SQL.
Yes, I tested things like:
- Unlocking jobs
- Late-join plus roundstart preferences
- Viewing player playtime in PP
- Viewing detailed player time report (linked in PP)
- Event fires regularly to add time
- Event fires for various types of EXP, adding correctly per department

Krausus commented:
Trying to track time spent actually playing specific jobs is one of those ideas that sound nice on paper, but turn out to be impractical to actually do properly. In this case...
- Changing your job doesn't give you time for the job you changed to.
- Changing your job *does* give you time for the job you changed from.
  - You can start as whatever job you want time on, then ask the HoP to reassign you to the job you actually want to play.
- Geneticists double-dip medical and science "experience", despite barely touching any of what is typically done in either of those departments.
- This system labels time spent as a job as "experience", regardless of whether you're actually getting any meaningful experience with the job. You could spend every shift getting wasted in the bar, locked up in the brig, sitting AFK in a cryopod, etc., and as long as you occasionally provide input, you'll still be earning experience.

The "wait X days to unlock job Y" system sucks, but this overcomplicates it with tracking that will need to be even more complicated to even get close to being accurate. And it *does* need to be accurate, or you *will* have players complaining when it isn't.

All we really need is to track time actually played, and use "played X hours" instead of "first connected X days ago" as the metric for what jobs are accessible. I'd bet very little code would actually have to be changed to accommodate that. Definitely less than 299 lines.

A couple other things:
- 2 DB queries, per player, every minute, all at the same time, will go very poorly on live. DB queries in BYOND are actually surprisingly expensive, and running 100-200 at once will probably freeze the server for a second or two, every minute.
- Copypasta. When you have 9 copies of a chunk of code, all nearly identical except for one or two things, you want to be iterating over a list that only specifies the differences, not copy-pasting.

Necaladun commented:
Re: Experience. - I think it has certain RPG connotations that mean it'd be better be named to playtime or something to make sure no one thinks they can spend it or level up. This is a stupidly minor point.

Re: Code based stuff like iterating over lists
I am not qualified to comment.

tl;dr
More logging and info good
More lag and code clutter bad
pls get me most logs and info with least lag and clutter

Krausus commented:
> I did consider making it account for job changes, but I decided not to, since...

Yes, it's not easy to do correctly, and any way you try to do it will have edge cases that mess things up. If we're to actually use this to track job play time, though, failing to account for the job-changing mechanics in the game is an issue.

> We can't measure how much you're learning on the job.

It's still labeled as experience rather than playtime, though. But that doesn't seem *quite* as bad now that I've realized players don't get a verb to view their experience.

> Don't ask "is it perfect". Ask "is this better than what we have now?" Yes, it is.

> This system is BETTER than what we have now.

Is it better than what we have now? Sure. The real question is, would it be better than a simpler system that only tracks time played, but does so with less code and more accurate results? That's not so simple to answer, and really depends on how accurate the job detection can be, and how badly we want to be able to see play time metrics.

VampyrBytes commented:
>  Copypasta. When you have 9 copies of a chunk of code, all nearly identical except for one or two things, you want to be iterating over a list that only specifies the differences, not copy-pasting.

> Examples, please.

What @Krausus means here would look something like this
```
var/list/positions = list("com" = command_positions, "sec" = security_positions, "sci" = science_positions) // etc
for(var/p in positions)
    if(C.mob.mind.assigned_role in positions[p])
          play_records[p] += minutes
          if(announce_changes)
               var/dept
               switch(p)
                   if("com")
                        dept = "Command"
                   if("sec")
                         dept = "Security"
                   if("sci")
                          dept = "Science"
              to_chat(C.mob,"<span class='notice'>You got: [minutes] [dept] EXP!")
```

Obviously with everything in the list and switch. Something like that anyway

VampyrBytes commented:
One thing, if you want to copy paste that, bear in mind that you can't put tabs in comments, so that's spaces, you'd need to sort that out

Kyep commented:
Cleanup Commit Changelog:
- 'EXP' is renamed to 'Playtime' in admin-visible commands/reports. System is still invisible to players.
- @Krausus and @VampyrBytes, changes have been made to reduce copied code
- @Necaladun, new config option "USE_EXP_RESTRICTIONS_HEADS_HOURS" added. When this is defined in config, and set to a >0 number, it overrides hours required for head jobs, allowing you to change the # of dept. hours required to unlock head jobs without having to change the code, or do PRs.

VampyrBytes commented:
Much better :smile: 
I did think about suggesting you change the com, sec etc to the full name to avoid having to have that switch in mine, but hadn't looked through to see what else that would impact, and didn't have time. Glad you figured it out as I really wasn't keen on that switch :wink:

VampyrBytes commented:
One little point, in code/helpers there's a file called global_lists.dm, `exp_jobsmap` should probably be defined in there as it is a handy list and could be reused elsewhere

Kyep commented:
@VampyrBytes global_lists.dm seems to be entirely made up of procs. Are you sure that's the file you were thinking of?

Krausus commented on SQL/paradise_schema.sql:
```diff
@@ -250,6 +250,7 @@ CREATE TABLE `player` (
   `nanoui_fancy` smallint(4) DEFAULT '1',
   `show_ghostitem_attack` smallint(4) DEFAULT '1',
   `lastchangelog` varchar(32) NOT NULL DEFAULT '0',
+  `exp` mediumtext NOT NULL DEFAULT '',
```
The `mediumtext` type cannot have a default value ([relevant page in MySQL reference manual](http://dev.mysql.com/doc/refman/5.7/en/blob.html)). You may want to use a varchar instead.

Krausus commented on SQL/paradise_schema_prefixed.sql:
```diff
@@ -250,6 +250,7 @@ CREATE TABLE `SS13_player` (
   `nanoui_fancy` smallint(4) DEFAULT '1',
   `show_ghostitem_attack` smallint(4) DEFAULT '1',
   `lastchangelog` varchar(32) NOT NULL DEFAULT '0',
+  `exp` mediumtext NOT NULL DEFAULT '',
```
Same as above.

Krausus commented on code/controllers/configuration.dm:
```diff
@@ -239,6 +244,18 @@
+					config.use_exp_restrictions_heads = 1
+
+				if("use_exp_restrictions_heads_hours")
+					config.use_exp_restrictions_heads_hours = value
```
This should `text2num` its value here, not in the code that uses it elsewhere.

Krausus replied:
Also, since this is used as minutes everywhere else, you should probably just convert this straight to minutes. The config option can stay hours, but you'll want to rename this var so it reflects that it's minutes.

tigercat2000 commented on code/game/jobs/job/engineering.dm:
```diff
@@ -17,7 +17,8 @@
 	minimal_player_age = 21
-
+	exp_requirements = 600
+	exp_type = "Engineering"
```
no magic strings please, use defines for all of these
`XP_TYPE_ENGINEERING` or something

Krausus commented on code/game/gamemodes/gameticker.dm:
```diff
@@ -67,6 +67,12 @@ var/round_start_time = 0
 		votetimer()
 
+/datum/controller/gameticker/proc/exptimer()
+	spawn(0)
```
The contents of this proc can be moved into a roundstart hook in `job_exp.dm`.

This spawn should match with the sleep below, or players will get 5 minutes of exp the moment a round starts.

5 should probably be in a define in `job_exp.dm`, after you've moved this there.

Krausus commented on code/game/gamemodes/gameticker.dm:
```diff
@@ -228,6 +234,8 @@ var/round_start_time = 0
 		spawn(3000)
 			statistic_cycle() // Polls population totals regularly and stores them in an SQL DB
+		if(config.use_exp_tracking)
+			exptimer()
```
This check would get rolled into the roundstart hook mentioned above.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,197 @@
+	set name = "Check Player Playtime"
+	if(!check_rights(R_ADMIN))
+		return
+	var/msg = "<html><head><title>Playtime Report</title></head><BR>Playtime:<BR>"
```
This HTML does not contain a body tag. It also does not close the html tag at any point below.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,197 @@
+		var/list/play_records = params2list(C.prefs.exp)
+		var/gen_exp_time = text2num(play_records["Crew"])
+		gen_exp_time = num2text(round(gen_exp_time / 60)) + "h"
+		msg += "<BR> - [key_name_admin(C.mob)]: <A href='?_src_=holder;getplaytimewindow=\ref[C.mob]'>" + C.get_exp_general() + "</a> "
```
It looks like you're using hyphens as makeshift bullet points. You can use `ul` and `li` tags to make lists with proper bullet points.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,197 @@
+		var/gen_exp_time = text2num(play_records["Crew"])
+		gen_exp_time = num2text(round(gen_exp_time / 60)) + "h"
+		msg += "<BR> - [key_name_admin(C.mob)]: <A href='?_src_=holder;getplaytimewindow=\ref[C.mob]'>" + C.get_exp_general() + "</a> "
+	if(msg != "")
```
`msg` starts with a default value, and only gets added to. This will never not be true.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,197 @@
+/client/proc/cmd_show_exp_panel(var/client/C in clients)
+	if(!check_rights(R_ADMIN))
+		return
+	var/body = "<html><head><title>Playtime for [C.key]</title></head><BR>Playtime:"
```
Another missing body tag, ironically in a var named `body`. The html tag *does* get closed below, though, so that's nice.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,197 @@
+
+	var/list/play_records = params2list(prefs.exp)
+	if(!play_records.len)
+		return "[src] has no records."
```
I'm not even sure what `[src]` outputs when used in a client proc. This should probably be `[key]`.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,197 @@
+			if(dep == "Exempt")
+				return_text += "<BR>- Exempt (all jobs auto-unlocked)"
+			else
+				return_text += "<BR>- [dep] " + get_exp_format(exp_data[dep]) + " (" + num2text(round(exp_data[dep]/exp_data["Crew"]*100)) + "%)"
```
At this point, there's no sanity check that played Crew time isn't 0. If it's 0 - for example, someone only has observer playtime so far - this will divide by 0 and runtime.

Also, if someone has very little crew playtime and a lot of observer playtime, I'm pretty sure this will show some massive %, since it seems to assume "Crew" encompasses all play time. There should probably be a category that increases any time any other category does.

Krausus replied:
Also, this whole proc uses numerous dashes-as-bullet-points. As previously mentioned, there are HTML tags for lists.

Also, you should really use text macros for these strings. `[get_exp_format(exp_data[dep])]` instead of `" + get_exp_format(exp_data[dep]) + "`

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,197 @@
+
+/client/proc/get_exp_format(var/expnum)
+	if(expnum > 0)
+		return num2text(round(expnum / 60)) + "h"
```
This will return 0h for anything under 60 minutes, due to `round` flooring numbers by default. You may want to simply include minutes in the text output here.

Kyep replied:
Thought about it, but decided against it, since the extra detail doesn't add anything. All numbers less than an hour equate to "very new" and never unlock anything.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,197 @@
+		return -1
+	spawn(0)
+		for(var/client/C in clients)
+			if(C.inactivity < (10 MINUTES))
```
You should change this to check for `>= (10 MINUTES)` and `continue` to accomplish what this currently does with one less level of indentation for the rest of the proc.

Krausus commented on code/modules/admin/topic.dm:
```diff
@@ -2016,6 +2016,12 @@
+		if(!check_rights(R_ADMIN))	return
+
+		var/mob/M = locate(href_list["getplaytimewindow"])
+		usr.client.cmd_show_exp_panel(M.client)
```
This doesn't sanity check to make sure `M` exists before trying to access its client. Stale topic links will make this runtime.

Krausus commented on code/modules/client/preference/preferences.dm:
```diff
@@ -551,6 +552,9 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 			HTML += "<del>[rank]</del></td><td><b> \[BANNED]</b></td></tr>"
 			continue
+		if(!has_exp_for_job(user, job.title))
+			HTML += "<del>[rank]</del></td><td> \[TIME]</td></tr>"
```
This should probably show the time and department. Like, [1h as Crew], [3h as Science], [10h as Command], etc.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,197 @@
+	var/body = "<html><head><title>Playtime for [C.key]</title></head><BR>Playtime:"
+	body += C.get_exp_report()
+	body += "</HTML>"
+	src << browse(body, "window=playerplaytime;size=550x615")
```
You may want to include the ckey in the window name here, so more than one playtime window can be open at a time for convenient comparison.

Kyep replied:
Not sure how to do this without breaking it. Also, not even Player Panel (the most opened panel for users) has this behavior currently.

Krausus replied:
`"window=playerplaytime;size=550x615"`
to
`"window=playerplaytime[C.ckey];size=550x615"`

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,197 @@
+		msg += "<BR> - [key_name_admin(C.mob)]: <A href='?_src_=holder;getplaytimewindow=\ref[C.mob]'>" + C.get_exp_general() + "</a> "
+	if(msg != "")
+		src << browse(msg, "window=Player_playtime_check")
+	log_admin("[key_name(usr)] checked player playtime")
```
I'm pretty sure looking at a list of players and their playtimes doesn't warrant an entry in the log.

VampyrBytes commented:
@Kyep 
Yeah, probably not (It was early in the morning, I wasn't properly awake :wink: ). Looking around, jobs.dm is probably the best place for it, around line 160, just after the other list defines. That way people looking at those lists will find it

Kyep commented:
@Krausus @VampyrBytes What do you think now?

Kyep commented:
@Krausus @VampyrBytes What do you think now?

alexpkeaton commented:
+1 to this. 

I think you should **add the capability to "dump" someone's EXP or immunity status, selectively or completely.**  To be utilized as an enforcement component separately and/or together with the current banning process.  For example, someone getting released from their second permaban in a month? Ok, but a second chance from the admins means you get your EXP slate wiped clean.  Particularly egregrious job ban? Ok, on release from it you have to earn that role again.  As a matter of procedure these sorts of acts should be relegated to either headmins or admin decision as a group, but the capability would be a very valuable tool to ensure compliance with rules.

SamHPurp commented:
I think this is an entirely logical system, and far advanced of our current one. With the appetite to remain listed 'indefinitely' - this seems like something we should already have, and I love everyone contributing to this PR.

VampyrBytes commented on code/game/jobs/job_exp.dm:
```diff
@@ -1,20 +1,22 @@
+#define EXP_TYPE_SPECIAL		"Special"
+#define EXP_TYPE_GHOST			"Ghost"
+#define EXP_TYPE_EXEMPT			"Exempt"
+
```
inb4 @tigercat2000 
Please sort out the spacing/tabs on these defines

tigercat2000 replied:
it probably works fine in tab size 4

Kyep replied:
It is aligned in DM.

VampyrBytes replied:
Ah right, just git being silly... did think it would be a bit strange to do them like this lol

Krausus replied:
These defines can't be here. They need to be in a file in the `__DEFINES` folder, or any code files that get read before this one won't be able to use them.

Krausus commented on code/game/jobs/job/security.dm:
```diff
@@ -60,6 +62,8 @@
 	minimal_access = list(access_security, access_sec_doors, access_brig, access_armory, access_court, access_maint_tunnels, access_weapons)
 	minimal_player_age = 21
+	exp_requirements = 300
+	exp_type = "Security"
```
Missed one.

VampyrBytes commented:
I don't do much html or sql, so I'll leave checking that over to @Krausus 
From my point of view it's looking much better, those defines make it much nicer (except the spacing on the one I commented on)
And it's a better placing for the list
:+1: :smile:

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,207 @@
+	if(!job_is_xp_locked(src.title))
+		return 0
+	if(check_rights(R_ADMIN, 0, C.mob))
+		return 1
```
With the change to what this proc returns, won't this lock all admins out of all jobs?

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,207 @@
+			if(dep == EXP_TYPE_EXEMPT)
+				return_text += "<LI>Exempt (all jobs auto-unlocked)</LI>"
+			else if(exp_data[EXP_TYPE_CREW] > 0)
+				return_text += "<LI>[dep] [get_exp_format(exp_data[dep])] (" + num2text(round(exp_data[dep]/exp_data[EXP_TYPE_CREW]*100)) + "%)</LI>"
```
Now that this checks for crew play time being above 0, someone with 0 minutes of crew play time won't show any play time stats at all, regardless of whether they have time as a ghost or non-crew antagonist.

I still think you should have another exp category that goes up any time any other category does, and use that instead of crew time to represent total time played.

Also, this still +es together strings. You never need to add together strings in a single line like this, just use embedded expressions. There's other lines that + together one-line strings, but I won't leave a comment on all of them because I am not a crazy person.

Also, you may want to separate particularly long things in strings, like the `num2text` call in this line, out into a separate var to improve the string's readability.

Kyep replied:
- You get crew playtime for playing any living mob. Even antags like nukeops. So you can't have antag exp but no crew exp.
- You can in theory ghost a lot and not have any crew playtime, but I'm changing this to it won't stop exp showing. Minor case though.
- +ing together strings is fixed.
- Having a separate 'time on server' category is pointless, EXP_TYPE_CREW + EXP_TYPE_GHOST together already equal this, and I'm already getting flak for complexity - I don't want to add more.

You get general playtime for *any* non-dead role. Including antags. So its not possible for someone to have antag exp but not crew exp. If you never actually join a round, but ghost a lot,

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,207 @@
+	for(var/client/C in clients)
+		var/list/play_records = params2list(C.prefs.exp)
+		var/gen_exp_time = text2num(play_records[EXP_TYPE_CREW])
+		gen_exp_time = num2text(round(gen_exp_time / 60)) + "h"
```
These two lines are now unused; you replaced them with `C.get_exp_general()`.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,207 @@
+	src << browse(msg, "window=Player_playtime_check")
+
+
+/client/proc/cmd_show_exp_panel(var/client/C in clients)
```
`in list` is for verbs, which this isn't. And please don't make it one. You can also drop the `var/`

This proc needs a sanity check to make sure `C` isn't null, since the topic that calls this does not do so.

This should probably be a `/datum/admins/proc`, so you can call it directly instead of using `usr.client.` in the topic that calls it. You'll have to change the `src <<` in this proc to `usr <<` when you change that.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,207 @@
+		var/list/play_records = params2list(C.prefs.exp)
+		var/gen_exp_time = text2num(play_records[EXP_TYPE_CREW])
+		gen_exp_time = num2text(round(gen_exp_time / 60)) + "h"
+		msg += "<LI> - [key_name_admin(C.mob)]: <A href='?_src_=holder;getplaytimewindow=\ref[C.mob]'>" + C.get_exp_general() + "</a></LI>"
```
... this topic uses a reference to the client's mob, which gets dereferenced back to the mob in the topic code, and then that mob's client gets used.

Just use a client reference here.

Krausus commented on code/modules/mob/new_player/new_player.dm:
```diff
@@ -246,6 +246,8 @@
 	if(!is_job_whitelisted(src, rank))	 return 0
 	if(!job.player_old_enough(src.client))	return 0
 	if(job.admin_only && !(check_rights(R_EVENT, 0))) return 0
+	if(job.available_in_playtime(src.client))
```
You don't need to use `src.client` here. Just use `client`. And yes, the line two up from this also uses an unnecessary `src.` - don't repeat their mistake.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,207 @@
+
+/hook/roundstart/proc/exptimer()
+	if(!config.sql_enabled || !config.use_exp_tracking)
+		return  0
```
Hook procs like this should `return 1`, or the hook handler will report an error. You should also `return 1` at the end of this proc.

Fox-McCloud commented:
This system comes across as hugely overcomplicated. A very simple system to merely track total play time (regardless of job) would be more than sufficient, IMO; it resolves the issue of "join once, come back 2 years later and griff with any job", but doesn't introduce 50 new things to maintain, bugfix, and attempt to resolve/integrate into the code.

The concept of "experience for a job" is flawed, IMO, because you can gain science experience by playing jobs other than science--likewise for medical doctor, etc. Furthermore, people can just request job changes and circumvent the system anyway---not to mention that there's plenty of people who will end up earning "experience" for one job despite getting a job change to another.

Just track total time played, IMO; it'll be easier to implement, maintain, and solves the ultimate problem that's being touched upon.

Spacemanspark commented:
@Fox-McCloud people can already ask the HoP and circumvent that restriction right now, and I'm pretty sure they can get in trouble for it if they get caught, too.

Fox-McCloud commented:
@Spacemanspark 

Uh, no. There are no restrictions--the only thing we ever remotely enforce is assigning karma jobs, and that's soft-enforced---having that for every single job would be an nightmare to account for in the code.

Kyep commented:
@Fox-McCloud 

Department-specific experience is only required for playing as the head of that department. E.g: security experience to play HoS. I think that makes total sense. I don't think that you can get, say, security experience from playing as any other department, and I don't think that someone with lots of hours of other jobs is necessarily qualified to play HoS without spending any time in security first.

HoP-based job transfers aren't the norm. People sticking with the department they signed up to play as, is. Allowing HoP-based job transfers to change your department for experience purposes probably requires allowing them to change your mind.assigned_role, which might have other implications in the code. It also seems more gamable than basing it on the job you signed up as.

tigercat2000 commented:
@Spacemanspark that's definitely not accurate, the only thing limited by admins is people evading jobbans or trying to take karma roles.

Spacemanspark commented:
I'm pretty sure I remember at least one person getting in trouble for doing it.

TullyBurnalot commented:
@Spacemanspark Highly unlikely. We do not prevent players from seeking transfers to jobs they have not yet unlocked. We only police jobbans and such as Tiger said.

Kyep commented:
@Krausus Take a look?

Necaladun commented:
I'd just like to re-iterate:

Nothing is it to be locked. This isn't a good place to discuss locking jobs. This is a system that may allow us to lock it differently in the future. Feel free to make forum posts etc about ideas of how to do so.

A tracking system like this I simply want atm to be able to check if people have an unreasonably high amount of antag/ghost time, or other outliers - as well as maybe be able to see what people's preferred departments are.

Options for locking are nice to have in place, but the emphasis on this PR should be just the tracking. Keep it atomic etc.

Krausus commented on code/modules/admin/topic.dm:
```diff
@@ -2016,6 +2016,13 @@
+		if(!check_rights(R_ADMIN))	return
+		var/mob/M = locate(href_list["getplaytimewindow"])
+		if(!M)
+			return
```
This needs feedback for when a mob can't be found, so admins clicking a non-working link know why. Though, again, it would be preferable if this topic used client references instead of mob references, since clients are what you're really dealing with.

Also, there's a one-line `if() return` above this. Those still exist in the codebase, but shouldn't be used in new code.

Kyep replied:
Feedback added.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,211 @@
+
+/datum/admins/proc/cmd_show_exp_panel(var/client/C)
+	if (!C)
+		return
```
This also needs feedback, instead of failing silently.

Kyep replied:
Feedback added.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,211 @@
+	var/body = "<html><head><title>Playtime for [C.key]</title></head><BODY><BR>Playtime:"
+	body += C.get_exp_report()
+	body += "</BODY></HTML>"
+	usr << browse(body, "window=playerplaytime;size=550x615")
```
Whoops, previous comment about this was in a collapsed thread. You can make this open separate windows like so:

`"window=playerplaytime[C.ckey];size=550x615"`

Or you could use `\ref[C]` instead of `[C.ckey]`. I'm not actually sure if there's some limit you'd hit with very long ckeys.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,211 @@
+				continue
+
+			var/DBQuery/exp_read = dbcon.NewQuery("SELECT exp FROM [format_table_name("player")] WHERE ckey='[C.ckey]'")
+			exp_read.Execute()
```
This doesn't check whether the query was successful. If this query fails, but the update query works fine, it will completely delete someone's `exp`.

Kyep replied:
Safety check added.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,211 @@
+	if(!establish_db_connection())
+		return -1
+	spawn(0)
+		for(var/client/C in clients)
```
Everything below here (except the sleep) should either be in a `try`/`catch` or another `spawn`, so if there's a runtime while trying to update exp for one client, later clients won't simply get skipped.

Instead of indenting all the code below again, you may want to simply move it to a new `/client/proc`, and have this proc call that one for each client.

Krausus commented:
> You get crew playtime for playing any living mob

Er... does *everything* have an `assigned_role`? Even...

`mind.assigned_role = "Corgi"`

... oh. So you get "crew exp" for playing a corgi. Or *any* simple animal. That doesn't really seem like an accurate label, then.

> Having a separate 'time on server' category is pointless, EXP_TYPE_CREW + EXP_TYPE_GHOST together already equal this, and I'm already getting flak for complexity - I don't want to add more.

Yeah... I didn't realize "crew" exp included time on literally any living mob, rather than, you know, actually playing as station crew. In that case, there should definitely be a separate category for total play time (counting what "crew" does now), and crew should only include pickable jobs.

As for complexity, you're already well past what's needed for better job gating. At this point, this is more about playtime metrics, which, honestly, I'd prefer a *more* complex tracking system for.

tigercat2000 commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,211 @@
+
+
+/datum/admins/proc/cmd_show_exp_panel(var/client/C)
+	if (!C)
```
:eyes: i see a space where there shouldn't be one

Kyep commented:
@Krausus Update 6 (just made) should resolve those things.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,198 @@
+				update_exp_client(L, mins, ann)
+			sleep(10)
+
+/proc/update_exp_client(var/client/C, var/minutes, var/announce_changes = 0)
```
Why not make this a `/client/proc`? That's basically what it is, but with the client passed in as `C` rather than being implicit.

Krausus commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,198 @@
+	var/hasread = 0
+	while(exp_read.NextRow())
+		read_records = params2list(exp_read.item[1])
+		hasread = 1
```
This... isn't quite what I meant. You can check the return value of `Execute()`, and if it returns false, it means the query failed. Usually gets used like this:
```
if(!query.Execute())
	var/err = query.ErrorMsg()
	log_game("SQL ERROR during something. Error : \[[err]\]\n")
	message_admins("SQL ERROR during something. Error : \[[err]\]\n")
	return
```
Though... I guess this should also work, since `NextRow` certainly won't return any rows if the query errors out. Hm.

Kyep replied:
Changed. Why not use both methods - one for a sql error, and one to detect whether the ckey exists.

Krausus commented on code/game/jobs/jobs.dm:
```diff
@@ -183,3 +183,19 @@ var/list/whitelisted_positions = list(
 
+var/global/list/exp_jobsmap = list(
+	EXP_TYPE_LIVING = list(), // all living mobs
+	EXP_TYPE_CREW = list(titles = command_positions + engineering_positions + medical_positions + science_positions + support_positions + supply_positions + security_positions + civilian_positions + whitelisted_positions), // crew positions
```
Oh geez. Do we really not have a list for all positions? This seems like something that should be one of the globals in `jobs.dm`.

Krausus replied:
Oh, this doesn't include the nonhuman positions. Might want to `|` these lists together to eliminate duplicate titles, though, in case anyone ever tries to iterate over them for some reason.

Kyep replied:
Nonhuman positions added.

Krausus commented:
Two things concern me about unlocking access to AI:
- It requires 5 hours of Silicon experience. Does cyborgification change your assigned role, or will people *only* get the experience if they happen to join as the one single cyborg job slot?
  - If they have to join as that one cyborg to ever be an AI, it will probably become the single hardest "head" job to unlock.
  - Otherwise, it might end up being the easiest, assuming they play cyborg properly and don't get silicon banned. That would probably be fine, though, since AI is one of the more heavily scrutinized jobs anyway.
- Maintenance drones get Cyborg as their assigned role, meaning they count towards unlocking AI. Probably want to override their `mind_initialize` to give them a different role, in `datums\mind.dm`.

> `ALTER TABLE player ADD exp mediumtext NOT NULL DEFAULT ''`

This needs to be updated to match the updated schema.

> Grandfather all active players, so they won't lose any jobs
> Be sure to enable the USE_EXP_TRACKING and USE_EXP_RESTRICTIONS_HEADS sections.

We probably shouldn't be turning on restrictions right away, and won't want to grandfather anyone. Once we begin tracking, active players will earn their time legitimately, and have it for when the restrictions turn on; the ones who don't weren't really active anyway.

Kyep commented:
@Krausus See updates in latest commit.
Also:
- Yes, cyborgification changes your assigned_role to 'Cyborg'.
- I do not mind drones / pAIs / etc counting as silicon EXP and thus giving credit towards AI.

Krausus commented:
> Yes, cyborgification changes your assigned_role to 'Cyborg'.

Ah, excellent.

> I do not mind drones / pAIs / etc counting as silicon EXP and thus giving credit towards AI.

Er... pAIs don't count toward Silicon time, and the fact that drones do is almost definitely an oversight that hasn't been an issue before now. Being able to bum around unnoticed as a drone for 5 hours and then go straight to AI definitely doesn't sound right.

Kyep commented:
https://github.com/ParadiseSS13/Paradise/blob/640746ee6b0d4388a8b793a49bd05091f5e9b941/code/game/jobs/jobs.dm
see: nonhuman_positions
pAIs DO count towards the nonhuman department. We can change that, but it might have other implications, so my PR currently doesn't.

Krausus commented:
`"AI", "Cyborg", "Drone", "pAI"`

... *what?!* I didn't even notice that, just that drones get "Cyborg" as their assigned role. This list gets used for roles in the jobban system, so it can't really be modified, but you probably shouldn't be using it, on account of it containing non-job roles.

Kyep commented:
@Krausus I can manually set the titles for EXP_TYPE_SILICON to "AI" and "Cyborg". Is this what you want?

Krausus commented:
Well, you wouldn't want them in the crew titles, either. But, yeah, that and making sure drones get the "Drone" assigned role they're supposed to have.

Kyep commented:
@Krausus There you go.

Iamgoofball commented:
One request

Don't out right ban the job

Just make players who aren't qualified super low priority for it

Kyep commented:
Sorry, won't be doing that in this PR. You can make one of your own that does it if you like.

Iamgoofball commented:
Gating is dumb and I will be doing a follow up PR to replace it

Krausus commented:
> A bugfix for Drone roles should really be a separate PR.

Not when it's going to be screwing up the tracking this PR adds.

> Gating is dumb and I will be doing a follow up PR to replace it

... we've been gating by account "age" for a long time. It's a little late to be getting up in arms about it.

tigercat2000 commented:
i don't think AI should require playtime as a borg
the two are insanely different roles, the only similarity is following laws properly
and that's
guess what
an admin matter, not a code matter

Kyep commented:
Update 7:
- Fix: Drones now get assigned_role = Drone.
- Tweak: AI is no longer considered a head job for playtime system configuration purposes.
- Tweak: Job gating (when enabled - its off by default) is all based on Crew playtime now. This is for all jobs, including heads and AI. Gating based on department playtime is still possible, but based on a separate config option that's also turned off by default.

KingPhilipIII commented:
I think AI might be

KingPhilipIII commented:
Goddamnit. Hit the wrong button on my phone(I can't delete the half-post). Pretend it was never here until I can delete it from my computer.

tigercat2000 commented on config/example/config.txt:
```diff
@@ -17,6 +17,19 @@ JOBS_HAVE_MINIMAL_ACCESS
+#USE_EXP_RESTRICTIONS_HEADS_DEPARTMENT 
+##Unhash this to enable playtime requirements for certain non-head jobs, like Engineer and Scientist.
+#USE_EXP_RESTRICTIONS_OTHER
+##Unhash this to allow admins to bypass job playtime requirements.
```
I don't think this is really necessary. Not every portion of a system needs to be a config option.

Kyep replied:
Options reworked since this comment. The majority of options were removed. This one in particular was left in because it is very useful for testing purposes.

Fox-McCloud commented:
This was discussed with @TheDZD @Earthdivine @Necaladun and myself.

I'm going to re-iterate what was stated to you earlier. Both maintainers and a head of staff have explicitly told you that extensive job-tracking is not something that is desired to be added nor something we inherently wish to have added. We are in agreement with a generalized play-time (per player) tracking system, but the other extensive parts of your system we have all agreed upon that we do not wish to be added in for now, until we can establish how well the basic system works.

Kyep commented:
@Necaladun @Fox-McCloud @TheDZD 

Update 8:
- Deleted department-based exp tracking
- Deleted exp definitions for all jobs except heads and AI
- Deleted the distinction between head and non-head jobs.
- Deleted most of the config options

Necaladun commented:
Yeah playtime tracking in general is useful for all sorts of things, great!

Krausus commented:
> Deleted exp definitions for all jobs except heads and AI

At this point, why have specific exp requirements on anything? You can just take `minimal_player_age`, multiply by X (60? 30?), and use that as a job's playtime requirement. This would let it pretty much directly replace the current age-based job gating.

Kyep commented:
@Krausus 
Reasons:
1) I want the numbers for this to be easy to tweak. Having them defined by a static formula prevents this.
2) I want the numbers to be informed by, but not dependent on, the age-based job gating system. Frankly the current age-based job gating system is terrible (for all the reasons outlined in the PR text) and I don't want this to be tied to it in any way.
3) What you suggest would give almost all jobs a playtime requirement. Should we ever turn on playtime requirements for jobs (it is OFF by default), I want us to be able to turn it on for only a subset of jobs (like heads/AI) first, and then look to review later. I do not want to make it impossible to apply it selectively by always copying settings from the age-based system.

Krausus commented:
> Having them defined on the job themselves makes them clear, consistent, and easily tweakable.

`minimal_player_age` is just as easy to tweak as a new var.

> having them defined by a static central formula makes them obscure, inconsistent with how other job properties are defined, and harder to tweak individually.

The formula would probably just be "instead of days old, it's hours played". And it would be no harder to tweak than a separate var would.

> The current age-based job gating system is terrible

Yes, that's why we want to replace it.

> and I don't want this to be tied to it in any way

You're just re-using a single var.

> we'll have to change this if we one day get rid of the age-gating system, so it could complicate things.

All we'll do is switch age-based gating off, and playtime-based gating on. If we decide to completely remove the age-based system, we would need to rename *a single var.*

> I want us to be able to turn it on for only a subset of jobs (like heads/AI) and then look to review the results before considering expanding it

Most jobs actually require very little time to unlock, and locking some players out of them would have less impact than locking players out of head roles. Either way, we'd want to wait until we're pretty confident anyone actively playing has accumulated a decent amount of time.

Kyep commented:
I still don't want to tie this to minimal_player_age, or the account age system in general.
I still do want to make it possible to have playtime-based gating for some jobs, but not all of the jobs currently gated via the account age system.
So, I'm not going to start using a formula based on minimal_player_age unless a maintainer or head of staff specifically asks me to.

TheDZD commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,192 @@
+// Procs
+
+
+/datum/job/proc/available_in_playtime(client/C)
```
Kind of a weird proc, since it returns 0 if the job *is* available, rather than 1.

Krausus replied:
This proc returns how much playtime until the job is available. 0 means "0 minutes until it's available".

TheDZD replied:
Welp, I read that entirely wrong then.

Kyep replied:
Two reasons why it is done this way:
- It is consistent with how player_old_enough/available_in_days already works.
- It enables use of the same function to see "is this person able to play this role" and "how many hours until they are?". Prevents copypaste.

Fox-McCloud commented:
According to TG apparently Yogstation already has a playtime tracking system that's up and running and working. May want to look into that.

Kyep commented:
I looked at Yogs' system, but I did not see anything in it I think is worth porting over to this.

Iamgoofball commented:
By a few hours how many hours exactly

Kyep commented:
@TheDZD Fixed merge conflict that arose recently from changes to admin_verbs.dm in another PR.

goofball: The description was out of date. I have updated it. Non-head jobs like sec officer have no requirements defined in this PR. The only jobs that do are head-level jobs. Even those requirements are disabled by default and have to be specifically enabled. If they are enabled, the requirement is 10h crew playtime. IE: 10 hours of playing any combination of crew jobs. Once this threshold is passed, all head jobs are unlocked together.

Markolie commented:
What's the plan for this PR at this stage?

Kyep commented:
My plan: wait for the maintainers to merge it.

Regen1 commented on SQL/paradise_schema.sql:
```diff
@@ -250,6 +250,7 @@ CREATE TABLE `player` (
   `nanoui_fancy` smallint(4) DEFAULT '1',
   `show_ghostitem_attack` smallint(4) DEFAULT '1',
   `lastchangelog` varchar(32) NOT NULL DEFAULT '0',
+  `exp` mediumtext,
```
What will this be? What will be inserted into this row, and why is it ´mediumtext´

Kyep replied:
Like "be_role" in the same table, "exp" is a comma-separated list of values. It tracks the number of minutes that the player has spent in each category for which stats are tracked. Here's an example value from my test database: "Living=610&Crew=610&Special=0&Ghost=30&Exempt=0". Translated, this means I've spent 10h10m as a living mob, 10h10m as a crew member, 30 minutes as a ghost, no time as a special role, and I'm not flagged Exempt. Records like that test record will be set for each active player, and it is 'mediumtext' because that's the type used for similar storage fields like 'be_role'.

Vivalas commented:
Wow is this the US government or Paradise? Way to see a great idea be watered down to nothing because of useless bureaucracy. Don't have even the slightest of the idea why the staff are so vehemently against job tracking based on department and role.

Kyep commented:
@Crazylemon64 All the changes you requested are done.

Crazylemon64 commented on code/game/jobs/job_exp.dm:
```diff
@@ -0,0 +1,192 @@
+				if(mob.mind.assigned_role in exp_jobsmap[cat]["titles"])
+					play_records[cat] += minutes
+					if(announce_changes)
+						to_chat(mob,"<span class='notice'>You got: [minutes] [cat] EXP!")
```
oh no, I accidentally got cat xp removed

time to commit sudoku 😢 
```
5 3 _ | _ 7 _ | _ _ _
6 _ _ | 1 9 5 | _ _ _
_ 9 8 | _ _ _ | _ 6 _
------+-------+------
8 _ _ | _ 6 _ | _ _ 3
4 _ _ | 8 _ 3 | _ _ 1
7 _ _ | _ 2 _ | _ _ 6
------+-------+------
_ 6 _ | _ _ _ | 2 8 _
_ _ _ | 4 1 9 | _ _ 5
_ _ _ | _ 8 _ | _ 7 9
```

Kyep commented:
Two small updates:
- Fixed a new merge conflict that arose after the Patreon PR was merged.
- Added an optional 5h playtime requirement to the officer/det/warden/borg jobs, to combat recent trend of people with no Paradise experience attempting to play security. @FreeStylaLT

Crazylemon64 commented:
Clean by me - nothing looks like it's become obsolete, either, so I'll go ahead and get the database prepared for the change, then merge this

tigercat2000 commented on config/example/config.txt:
```diff
@@ -17,6 +17,13 @@ JOBS_HAVE_MINIMAL_ACCESS
+##Unhash this to enable playtime requirements for jobs that have them defined.
+#USE_EXP_RESTRICTIONS
+##Unhash this to allow admins to bypass job playtime requirements.
+#USE_EXP_RESTRICTIONS_ADMIN_BYPASS
```
This should either be
A) Uncommented by default
B) Changed to be an option of "DISABLE_EXP_RESTRICTION_ADMIN_BYPASS", and make the bypass on by default in the configuration controller.

Kyep replied:
Changed
